### ISSUE TITLE
missing condesc_t in dict-structures.h

### ISSUE THREAD

**Issue by mjsduncan** [opencog/link-grammar#710]
this is on a fresh install of ubuntu 17.10 and link-grammar 5.4.4 tarball
```
[ 49%] Building CXX object opencog/nlp/lg-dict/CMakeFiles/lg-dict-entry.dir/LGDictExpContainer.cc.o
In file included from /usr/local/include/link-grammar/dict-api.h:17:0,
                 from /home/mozi/oc/opencog/opencog/nlp/lg-dict/LGDictExpContainer.h:27,
                 from /home/mozi/oc/opencog/opencog/nlp/lg-dict/LGDictExpContainer.cc:27:
/usr/local/include/link-grammar/dict-structures.h:18:10: fatal error: connectors.h: No such file or directory
 #include "connectors.h"
          ^~~~~~~~~~~~~~
```

**Comment by ampli** [comment]
I added it so editing `dict-structures.h` with an IDE will not flag errors due to the recently added `condesc_t` struct member of Exp_struct. But when used from `LGDictExpContainer.h`, it is not found since it is not installed in the system.

There is a worse problem:
In `LGDictExpContainer.cc`, `exp->u.string` is used. But now the Exp struct doesn't contain this field.

@linas, please advise...

**Comment by ampli** [comment]
As a work around, try to use link-grammar 5.4.3 instead of 5.4.4.

**Comment by linas** [comment]
I will fix `exp->u.string`  ...   mostly just have to move `conndesc_t` to `dict-structures.h`

**Comment by ampli** [comment]
Instead, I can implement #577.
I will continue there.

**Comment by linas** [comment]
Amir -- both the pull reqs #711 and #715 provide fixes for this bug. Can you look and see which is better? Or if there is a third way?

I don't like #711 because it puts "private" things into the public API.
For this reason, #715 was supposed to be better, except that now there is a string pointer in both `struct Conn` and in `condesc_t`, which is kind-of-ugly.  

Is there a better way?

**Comment by ampli** [comment]
Readding the string to the connector struct doesn't seem "right", due to duplicate information
and making the struct bigger (on the other hand it makes the access to the string faster).

I will try to propose another way, but I need to understand the following:

What is wrong with adding a public API `const char* get_expression_string(Exp*)`, leaving `condesc*` as an opaque pointer?
Or alternatively, if needed, `const char* get_connector_string(condesc_t*)` that can be used on the corresponding Exp and Connector fields (again leaving it as an opaque implementation-dependent pointer)?
Is it not enough?

**Comment by linas** [comment]
Closing; there is a work-around for now, and a fix will be published in link-grammar version 5.5.0, which will be out "real soon now"
